quinn-proto/src/connection.rs

```diff
@@ -813,43 +809,37 @@ impl Connection {
                 }
-                let state = self.state.take().unwrap();
-                self.handle_connected_inner(mux, now, number, packet, state)
+                self.handle_connected_inner(mux, now, number, packet)
             }
         };
 
         // State transitions for error cases
-        let state = match result {
-            Ok(state) => state,
-            Err(conn_err) => {
-                mux.emit(conn_err.clone().into());
-
-                match conn_err {
-                    ConnectionError::ApplicationClosed { reason } => State::closed(reason),
-                    ConnectionError::ConnectionClosed { reason } => State::closed(reason),
-                    ConnectionError::Reset => {
-                        debug!(self.log, "unexpected connection reset error received"; "err" => %conn_err, "initial_conn_id" => %self.init_cid);
-                        panic!("unexpected connection reset error received");
-                    }
-                    ConnectionError::TimedOut => {
-                        debug!(self.log, "unexpected connection timed out error received"; "err" => %conn_err, "initial_conn_id" => %self.init_cid);
-                        panic!("unexpected connection timed out error received");
-                    }
-                    ConnectionError::TransportError { error_code } => State::closed(error_code),
-                    ConnectionError::VersionMismatch => State::Draining,
+        if let Err(conn_err) = result {
+            mux.emit(conn_err.clone().into());
+            self.state = match conn_err {
+                ConnectionError::ApplicationClosed { reason } => State::closed(reason),
+                ConnectionError::ConnectionClosed { reason } => State::closed(reason),
+                ConnectionError::Reset => {
+                    debug!(self.log, "unexpected connection reset error received"; "err" => %conn_err, "initial_conn_id" => %self.init_cid);
+                    panic!("unexpected connection reset error received");
                 }
-            }
-        };
+                ConnectionError::TimedOut => {
+                    debug!(self.log, "unexpected connection timed out error received"; "err" => %conn_err, "initial_conn_id" => %self.init_cid);
+                    panic!("unexpected connection timed out error received");
+                }
+                ConnectionError::TransportError { error_code } => State::closed(error_code),
+                ConnectionError::VersionMismatch => State::Draining,
+            };
+        }
 
-        if !was_closed && state.is_closed() {
+        if !was_closed && self.state.is_closed() {
             self.close_common(mux, now);
         }
 
         // Transmit CONNECTION_CLOSE if necessary
-        if let State::Closed(ref state) = state {
-            self.transmit_close(mux, now, &state.reason);
+        if let State::Closed(ref state) = self.state {
+            self.transmit_close(mux, now, &state.clone().reason);
```

Comment:
If we modify `Connection::close` to update the state *before* calling `transmit_close`, we could in turn get rid of the argument to `transmit_close` and have it read from `self.state` directly, avoiding this clone.

Reply:
Let's do that as a follow-up, since it is at least moderately finicky to get right.